rocks-lib/src/operations/install.rs

```diff
@@ -54,10 +60,41 @@ async fn install_impl(
         .expect("could not find rockspec in source directory. this is a bug, please report it.")
         .into_path();
 
-    crate::build::build(
-        progress,
-        Rockspec::new(&std::fs::read_to_string(rockspec_path)?)?,
-        config,
-    )
-    .await
+    let rockspec = Rockspec::new(&std::fs::read_to_string(rockspec_path)?)?;
+
+    // TODO(vhyrro): Create a unified way of accessing the Lua version with centralized error
+    // handling.
+    let lua_version = rockspec.lua_version();
+    let lua_version = config.lua_version().or(lua_version.as_ref()).ok_or_eyre(
+        "lua version not set! Please provide a version through `--lua-version <ver>`",
+    )?;
+
+    let tree = Tree::new(config.tree().clone(), lua_version.clone())?;
+    let mut lockfile = tree.lockfile()?;
+
+    let package = lockfile.add(
+        &LuaPackage::new(rockspec.package.clone(), rockspec.version.clone()),
+        LockConstraint::Constrained(package_req.version_req().clone()),
+        false,
+    );
+
+    // Recursively build all dependencies.
+    // TODO: Handle regular dependencies as well.
+    for dependency_req in rockspec
+        .build_dependencies
+        .current_platform()
+        .iter()
+        .filter(|req| tree.has_rock(req).is_none())
```

Comment:
Important note: we will likely want to remove this *and* also encode constraint information in our UID system.

A dependency that is bound to `>= 1.0.0` is different to a dependency that is bound to `<= 1.0.0`. This filter command doesn't install a dependency if one already exists and also satisfies the version constraint.

This will cause problems in the following case, however:
```
| rock1
|- dependency 1.0.0 (<= 1.0.0)
| rock2
|- dependency 1.0.0 (>= 1.0.0)
```

In the following scenario, both dependencies will share the same directory,. During an update, however, `rocks update` will read the lockfile and see only a single constraint (maybe `>= 1.0.0` for example), and will happily upgrade `dependency` to a newer version, breaking the constraints of `rock1`.

I have considered simply finding the convergence point of both constraints and mixing them in the lockfile (`dependency <= 1.0.0, dependency >= 1.0.0`, functionally equivalent to `dependency == 1.0.0`), but then `rock2` will never benefit from a newer version of `dependency`, since it'll be artificially locked by the constraints of `rock1`.

If we encode the constraint in the UID hash as well, then these two dependencies will be fundamentally different. I'm starting to think that in our rocktrees we should start encoding hashes in the directory names, just like the Nix store (`hash-package@version` instead of just `package@version`).

Reply:
Since the latest commit the ID now also contains constraint information. I'll create a stacked PR on top of this to encode hashes in the rocktree too.

Reply:
#119 was a necessity before implementing rocktree hashes, it should be fairly easy from here.

Reply:
see #120